### PR TITLE
[refact] Move components computation to common class (generators)

### PR DESCRIPTION
**Pull request by jgsogo**
Changelog: omit
Docs: omit

Some logic about components can be shared between `cmake_find_package[_multi]` and `pkg_config` generators.

This PR makes it easier to understand changes in https://github.com/conan-io/conan/pull/7644

#TAGS: slow

### CODE REVIEW

File: conans/model/conan_generator.py [hunk 1]
```diff
@@ -43,5 +51,48 @@ def content(self):
     def filename(self):
         raise NotImplementedError()
 
+
+class GeneratorComponentsMixin(object):
```
Comment (danimtb) [inline conans/model/conan_generator.py:55]: why not adding this to the `Generator` class?

**Review by danimtb — APPROVED** [review]

File: conans/client/generators/cmake_find_package.py [hunk 2]
```diff
@@ -229,38 +223,16 @@ def content(self):
...
-        find_package_components.reverse()  # From the less dependent to most one
-        return find_package_components
-
-    def _get_component_requires(self, pkg_name, pkg_findname, comp):
-        comp_requires_findnames = []
-        for require in comp.requires:
-            if COMPONENT_SCOPE in require:
-                comp_require_pkg_name, comp_require_comp_name = require.split(COMPONENT_SCOPE)
-                comp_require_pkg = self.deps_build_info[comp_require_pkg_name]
-                comp_require_pkg_findname = self._get_name(comp_require_pkg)
```
Comment (memsharded) [inline conans/client/generators/cmake_find_package.py:248]: Is it possible that this refactor hides a bug?
- In the previous implementation, the ``comp_require_pkg_findname`` is different when it is in another package. 
- The new implementation seems to use always the same.

Can this behavior be checked with a test to show that it was a bug in the past? Or am I missing something? 
I am finding this refactor is a bit difficult to understand while reading it that it is not changing behavior.

File: conans/model/conan_generator.py (hunk 1 above)
Comment (jgsogo) [reply conans/model/conan_generator.py:55]: Just to keep things isolated. If you have the components feature (`cmake_find_package` and `pkg_config`) you add this class to your inheritance and then you get some new powers/overrides for your functions. In the next PR you  can see an example here: https://github.com/conan-io/conan/pull/7644/files#diff-496c84b81260eff9ea669a196e9054dbR104, you get the new implementation of the `get_public_deps` only if you add the mixin to your parent classes. 

Usually (not the best example here) this avoids and extra `if/else` and reduces the [complexity](https://en.wikipedia.org/wiki/Cyclomatic_complexity) of the function and the number of unittests required (type/inheritance is hardcoded at _compile time_, while `if/else` is resolved at runtime).

File: conans/client/generators/cmake_find_package.py (hunk 2 above)
Comment (jgsogo) [reply conans/client/generators/cmake_find_package.py:248]: The new implementation takes into account if the component comes from a different package or from the same one:

```python
pkg, cmp = req.split(COMPONENT_SCOPE) if COMPONENT_SCOPE in req else (pkg_name, req)
```

The `pkg` variable uses the one from the requires (if `req` is something like `pkg::cmp`) or the one from the same package (if `req` is something like `cmp`).

It was not a bug and it is not a bug. We need to use the package that provides the component and call `_get_name`  with it. IMO it doesn't require any extra test, there are already lot of tests using components and it is very hard to have a false positive here: it'd require a package with that name in `deps_build_info` providing a component `deps_build_info[pkg_name].components[cmp_name]` and returning the expected value.

**Review by memsharded — APPROVED** [review]

**Pull request merged**